File: Submissions/930_Binary_Subarrays_With_Sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& A, int S) {
        int result = 0;
        map<int, int> count;
        int sum = 0;
        
        int n = A.size();
        for(int i=0; i<n; ++i) {
            sum += A[i];
            if(sum - S >= 0) result += count[sum-S];
            if(sum == S) ++result;
            ++count[sum];
        }
        
        return result;
    }
};

File: Kotlin/930_Binary_Subarrays_With_Sum.kt
class Solution {
    fun numSubarraysWithSum(A: IntArray, S: Int): Int {
        var result = 0
        val count = hashMapOf<Int, Int>()
        var sum = 0
        
        for(i in 0 until A.size) {
            sum += A[i]
            if(sum - S >= 0) result += count.getOrDefault(sum - S, 0)
            if(sum == S) ++result
            count[sum] = count.getOrDefault(sum, 0) + 1
        }
        
        return result
    }
}